content/api/servo-assembly/arduino-commands.md

```diff
@@ -0,0 +1,196 @@
+returned as part of the message if you call the `help` command.
+
+## Calling commands
+
+To call a command from Python you should call the `direct_command` method on a
+`Servo` instance. You can send strings and numbers:
+
+```python
+r.servo_board.direct_command('echo', "first argument", 2, "third argument")
+```
+
+This will return a `list` of the responses from the command (anything which was
+sent back using the `serialWrite` function):
+
+```python
+>>> r.servo_board.direct_command('echo', "first argument", 2, "third argument")
+["first argument", "2", "third argument"]
+```
+
+Note that the responses will all be strings, so you'll need to explicitly
+convert them back to numbers if you want to use them as numbers:
+
+``` python
+>>> a, b = r.servo_board.direct_command('echo', 6, 2)
+>>> print(int(a) / int(b))
+3
+>>> print(a / b)
+Traceback (most recent call last):
+  File "<stdin>", line 1, in <module>
+TypeError: unsupported operand type(s) for /: 'str' and 'str'
+```
+
+### Errors
+
+If an error is encountered while handling a command an [exception][errors] will
+be raised. The exceptions will contain information about the error so that you
+can change your code to fix it.
+
+#### `CommandError`
+
+A `CommandError` indicates that the servo assembly experienced an error in
+processing the command and sent back an error message. The error message will be
+that returned by the command on the Arduino as a `COMMAND_ERROR`:
+
+```python
+>>> r.servo_board.direct_command('echo')
+Traceback (most recent call last):
+  File "<stdin>", line 1, in <module>
+CommandError: Error: nothing to echo
+```
+
+You'll also see a `CommandError` if you try to call a command which doesn't
+exist:
+
+```python
+>>> r.servo_board.direct_command('missing-command')
+Traceback (most recent call last):
+  File "<stdin>", line 1, in <module>
+CommandError: Error, unknown command: missing-command
+```
+
+This error will generally indicate that something is wrong in your Python code.
+
+#### `InvalidResponse`
+
+A `InvalidResponse` indicates that the Python API which handles the actual
+communication to the Arduino received an invalid response from the Arduino. The
+error message will be the actual text received from the Arduino, allowing you to
+easily see the cause of the problem.
+
+```python
+>>> r.servo_board.direct_command('the-command')
+Traceback (most recent call last):
+  File "<stdin>", line 1, in <module>
+InvalidResponse: b'received data'
+```
+
+This error will generally indicate that one of the messages being sent back from
+the Arduino is badly formatted. Check to ensure that you're using `serialWrite`
+correctly for all your messages back to the Pi.
+
+
+<!--
+    At some point we might need to detail how to remove bits of the firmware
+    which the teams aren't using. That's intentionally omitted for the moment
+    as it's nontrivial to describe which bits are safe to remove.
+-->
```

Comment:
Haven't tested this but I would imagine that removing or commenting an entry in the `commands` away should cause the corresponding handler function to be optimised out of the binary. This should produce some initial space savings, at least.

Reply:
This does indeed seem to happen. I think I'd still rather defer this for now.

Reply:
Yeah, that's something for separate PR.